subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -351,9 +346,9 @@ public DefaultConfiguration(
         this.path = domainObjectContext.projectPath(name);
         this.defaultConfigurationFactory = defaultConfigurationFactory;
 
-        this.canBeConsumed = roleAtCreation.isConsumable();
-        this.canBeResolved = roleAtCreation.isResolvable();
-        this.canBeDeclaredAgainst = roleAtCreation.isDeclarableAgainst();
+        this.allowedUsage = Usage.CONSUMABLE.toggle(roleAtCreation.isConsumable(), this.allowedUsage);
```

Comment:
Having these methods on `Usage` makes using this seem clunky. A more user friendly API would look more like

```
allowedUsage.setEnabled(Usage.CONSUMABLE, roleAtCreation.isConsumable())
```

This could easily be accomplished by having `allowedUsage` being an instance of some `Role` class to encapsulate the `byte` state.

Reply:
Replacing the byte field with an Object wouldn't trim as much memory from the state saved by `DefaultConfiguration`, which was the goal.  Plus it's a bigger change than I want to make here.  With the previous change + static imports, it looks like:

`this.allowedUsage = CONSUMABLE.updateUsage(false, this.allowedUsage);` 

which I'll admit isn't quite as good as your example, but I think is better and clear enough for something temporary like this.

Reply:
How temporary is this? I agree adding the object reference takes away most of the advantage of this change. I question whether this change actually makes a significant dent in the memory usage or if we're just playing with pennies at this point

Reply:
It's temporary until the usage is permanently locked to roles.  Might be possible prior to splitting implementation classes, but that's still TBD.  Hopefully gone no later than 9.0.

It's pennies for sure, I mentioned in the description it was only for marginal gains.  I figured it was be easier to make the change to check for any performance impact from this and then discuss in the review whether we wanted to do it or not.  I lean towards including it, but I'm not heavily invested here.